File: src/main/scala/com/github/nichijou/algorithms/MergeSort.scala
package com.github.nichijou.algorithms

import com.github.nichijou.utils.RandomArr

/**
  * Created by zk_chs on 16/4/27.
  * 归并排序,自顶向下,所需时间与NlogN成正比,缺点为所需的额外空间也与N成正比
  */
object MergeSort extends Example[Char] {

  var aux: Array[Char] = _

  override def sort(a: Array[Char]) = {
    aux = new Array[Char](length)
    sort(a, 0, length - 1)
    show(a)
  }

  def sort(a: Array[Char], lo: Int, hi: Int): Unit = {
    if (hi <= lo) return
    val mid = lo + (hi - lo) / 2
    if (a.length > 15) {
      sort(a, lo, mid)
      sort(a, mid + 1, hi)
      merge(a, lo, mid, hi)
    } else
      InsertionSort.sort(a) // 使用插排处理小的数组

  }

  def merge(a: Array[Char], lo: Int, mid: Int, hi: Int) = {
    var i = lo
    var j = mid + 1
    for (k <- lo to hi) aux(k) = a(k)
    for (k <- lo to hi)
      if        (i > mid)               {a(k) = aux(j); j += 1}
      else if   (j > hi)                {a(k) = aux(i); i += 1}
      else if   (less(aux(j), aux(i)))  {a(k) = aux(j); j += 1}
      else                              {a(k) = aux(i); i += 1}
  }

//  val length = 20
//  val arr = RandomArr.shuffle(length)
//  println(arr.mkChar(" "))
//  sort(arr)
  val s = "asmnqoakqwnkdksfpqpowqenvsjhsakewqncammxnckjehuiuqwnejkasjncsjakbdqwijncajxkajdkqjwhejnjkxcnajkfbocviuwetiyu"
  val length = s.length
  println(s.toArray.mkString(" "))
  val start = System.currentTimeMillis()
  sort(s.toArray)
  val end = System.currentTimeMillis()
  println("total: " + (end - start))

}

File: src/main/java/com/github/algorithms/MergeSort.java
package com.github.algorithms;

import java.util.Arrays;

/**
 * Created by zk_chs on 16/4/28.
 */
public class MergeSort extends Example {

    private Comparable[] aux;

    @Override
    void sort(Comparable[] arr) {
        aux = new Comparable[arr.length];
        sort(arr, 0, arr.length - 1);
    }

    private void sort (Comparable[] arr, int lo, int hi){
        if (lo >= hi) return;
        int mid = lo + (hi - lo) / 2;
        sort(arr, lo, mid);
        sort(arr, mid + 1, hi);
        merge(arr, lo, mid, hi);
    }

    private void merge (Comparable[] arr, int lo, int mid, int hi){
        int i = lo; int j = mid + 1;
        for (int k = lo; k <= hi; k++) aux[k] = arr[k];
        for (int k = lo; k <= hi; k++)
            if (i > mid)                    arr[k] = aux[j++];
            else if (j > hi)                arr[k] = aux[i++];
            else if (less(aux[j], aux[i]))  arr[k] = aux[j++];
            else                            arr[k] = aux[i++];
    }

    public static void main(String[] args) {
        Integer[] arr = RandomArr.shuffle(30);
        System.out.println(Arrays.toString(arr));
        MergeSort sort = new MergeSort();
        long start = System.nanoTime();
        sort.sort(arr);
        long end = System.nanoTime();
        System.out.println(Arrays.toString(arr));
        System.out.println("total: " + (end - start));
    }

}
